compiler/src/dmd/typesem.d

```diff
@@ -2994,7 +2994,7 @@ Type typeSemantic(Type type, Loc loc, Scope* sc)
         {
             /* struct S s, *p;
              */
-            return returnType(mtype.resolved.addSTC(mtype.mod));
+            return returnType(mtype.resolved.addSTC(cast(STC) mtype.mod));
```

Comment:
Case in point: passing `mtype.mod` to `addSTC` looks like a bug, but since this is a refactoring PR, I'll change that in a next PR.